### PR TITLE
feat: clean procedure's state after it is done

### PR DESCRIPTION
**Pull request by evenyag**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?
This PR cleans the procedure's state in `ProcedureStore` and memory after it is done
- Add `delete_procedure()` method to `ProcedureStore`
- After a root procedure is done, the runner cleans all the state files related to this procedure and its children
- Removes the metadata in memory after `META_TTL`. When we submit a procedure to the manager or query a procedure's state, the manager tries to clean the outdated metadata
- Cleans finished procedures during recovery

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
- #1509

### CODE REVIEW

File: src/common/procedure/src/local.rs [hunk 1]
```diff
@@ -429,7 +429,7 @@ impl ProcedureManager for LocalManager {
         logging::info!("LocalManager start to recover");
 
         let procedure_store = ProcedureStore::new(self.state_store.clone());
-        let messages = procedure_store.load_messages().await?;
+        let (messages, finished_ids) = procedure_store.load_messages().await?;
```
Comment (killme2008) [inline src/common/procedure/src/local.rs:432]: Suggestion: Add some metrics for procedure recovery or record the cost time in the final log.

File: src/common/procedure/src/local.rs [hunk 2]
```diff
@@ -412,14 +457,33 @@ impl ProcedureManager for LocalManager {
...
     }
 
     async fn procedure_state(&self, procedure_id: ProcedureId) -> Result<Option<ProcedureState>> {
+        self.manager_ctx.remove_outdated_meta(META_TTL);
+
         Ok(self.manager_ctx.state(procedure_id))
     }
 
     fn procedure_watcher(&self, procedure_id: ProcedureId) -> Option<Watcher> {
+        self.manager_ctx.remove_outdated_meta(META_TTL);
```
Comment (killme2008) [inline src/common/procedure/src/local.rs:485]: Why not use a `RepeatedTask` to purge outdated states periodically? Then we don't have to add so many hooks in code.

File: src/common/procedure/src/store.rs [hunk 3]
```diff
@@ -119,9 +127,50 @@ impl ProcedureStore {
...
+    pub(crate) async fn delete_procedure(&self, procedure_id: ProcedureId) -> Result<()> {
+        let path = proc_path!("{procedure_id}/");
+        // TODO(yingwen): We can optimize this to avoid reading the value.
+        let mut key_values = self.0.walk_top_down(&path).await?;
+        // 8 should be enough for most procedures.
+        let mut step_keys = Vec::with_capacity(8);
+        let mut finish_keys = Vec::new();
+        while let Some((key, _)) = key_values.try_next().await? {
+            let Some(curr_key) = ParsedKey::parse_str(&key) else {
+                logging::warn!("Unknown key while deleting procedures, key: {}", key);
```
Comment (killme2008) [inline src/common/procedure/src/store.rs:140]: adds the procedure id to log.

Comment (evenyag) [reply src/common/procedure/src/store.rs:140]: The key already contains the procedure id.

File: src/common/procedure/src/local.rs (hunk 2 above)
Comment (evenyag) [reply src/common/procedure/src/local.rs:485]: I just don't want to implement the TaskFunction trait. I'll fix it.

File: src/common/procedure/src/local.rs [hunk 4]
```diff
@@ -126,6 +129,8 @@ pub(crate) struct ManagerContext {
     procedures: RwLock<HashMap<ProcedureId, ProcedureMetaRef>>,
     /// Messages loaded from the procedure store.
     messages: Mutex<HashMap<ProcedureId, ProcedureMessage>>,
+    /// Ids and finished time of finished procedures.
+    finished_procedures: RwLock<Vec<(ProcedureId, Instant)>>,
```
Comment (MichaelScofield) [inline src/common/procedure/src/local.rs:133]: Will `finished_procedures` keep growing, consuming too many memory?

File: src/common/procedure/src/local.rs (hunk 2 above)
Comment (evenyag) [reply src/common/procedure/src/local.rs:485]: Using `RepeatedTask` requires adding `start()` and `stop()` to the procedure manager.

File: src/common/procedure/src/local.rs (hunk 4 above)
Comment (evenyag) [reply src/common/procedure/src/local.rs:133]: Yes, I have noticed this issue and replaced it with a VecDeque.

File: src/common/procedure/src/local.rs (hunk 2 above)
Comment (evenyag) [reply src/common/procedure/src/local.rs:485]: I plan to leave this as a good first issue. I realize that the `RepeatedTask` doesn't need an async start function. I'd refactor it in another PR.

**Review by killme2008 — APPROVED** [review]
LGTM

**Review by MichaelScofield — APPROVED** [review]

**Pull request merged**